### PR TITLE
Fix threading bug when multiple destroy_key calls run on the same key

### PR DESCRIPTION
**Pull request by Ryan-Everett-arm**
## Description
There is a period of time in `psa_destroy_key` between registering to read a key slot and locking the mutex to actually do the destruction. It is possible that during this period a rival thread can complete a `psa_destroy_key` call on the same key.
Previously, this would result in both callers returning `PSA_SUCCESS` (one would goto exit having failed the state transition, and then have status overwritten by `PSA_SUCCESS` in the exit logic, where status is set to `psa_unregister_read`).

This PR changes this so that exactly one thread will return `PSA_SUCCESS`. Double destruction is not behaviour defined by the PSA spec, but doing the right thing here allows us to have better testing. This fix lets us run multi-threaded tests in which we can have many threads operating and destroying the same key, asserting that no corruption occurs and exactly one thread succeeds.  

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [x] **backport** not required
- [ ] **testing** this bug was picked up by a test I am currently writing, which will be uploaded soon.

### CODE REVIEW

File: library/psa_crypto.c [hunk 1]
```diff
@@ -1106,6 +1106,17 @@ psa_status_t psa_destroy_key(mbedtls_svc_key_id_t key)
...
+    if (slot->state == PSA_SLOT_PENDING_DELETION) {
+        /* Another thread has destroyed the key between us locking the slot
+         * and us gaining the mutex. Unregister from the slot,
+         * and report that the key does not exist. */
+        status = psa_unregister_read(slot);
+
+        PSA_THREADING_CHK_RET(mbedtls_mutex_unlock(
+                                  &mbedtls_threading_key_slot_mutex));
+        return (status == PSA_SUCCESS) ? PSA_ERROR_INVALID_HANDLE : status;
+    }
```
Comment (tom-cosgrove-arm) [inline library/psa_crypto.c:1119]: Any way of adding a test for this?

Comment (paul-elliott-arm) [reply library/psa_crypto.c:1119]: Not entirely sure if I am interpreting this correctly, but the tests that found this issue will come in a seperate PR this week.

Comment (Ryan-Everett-arm) [reply library/psa_crypto.c:1119]: That is the plan. The test which catches this also tests concurrently using the same key in a myriad of ways. Doing this requires some overhaul to `psa_exercise_key.c` which I am currently working on. So my current plan is to put that in a different PR

**Review by paul-elliott-arm — APPROVED** [review]
I think this LGTM.

I have been trying to think of other areas where we could have similar issues / other states that this could happen with, but so far I haven't come up with anything.

File: library/psa_crypto.c (hunk 1 above)
Comment (tom-cosgrove-arm) [reply library/psa_crypto.c:1119]: Ah, okay, thanks for the reassurance

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by Ryan-Everett-arm** [comment]
This is now being tested via #8924

**Pull request merged**